File: 2111-70-1910-remove-all-occurrences-of-a-substring/2111-70-1910-remove-all-occurrences-of-a-substring.cpp
class Solution {
public:
    short lps[1001];
    void kmp(string part){
        int j = 0;
        for(int i = 1 ; i < part.size();i++){
            while(j && part[i] != part[j]) j = lps[j-1];
            if(part[i] == part[j])lps[i] = ++j;
        }
    }
    string removeOccurrences(string s, string part) {
        int sz = part.size();
        kmp(part);
        vector<short>stk;
        string ans;
        int j = 0;
        for(char ch : s){
            ans.push_back(ch);
            stk.push_back(j);
            while(j && ch != part[j]) j = lps[j-1];
            if(ch == part[j]) ++j;
            stk.back() = j;
            if( j == sz){
                ans.resize(ans.size()-sz);
                stk.resize(stk.size()-sz);
                j = stk.size() ? stk.back() : 0 ;
            }
            
        }
        return ans;


    }
};

File: 2111-70-1910-remove-all-occurrences-of-a-substring/2111-70-1910-remove-all-occurrences-of-a-substring.java
class Solution {
    public String removeOccurrences(String s, String part) {
        int sz = part.length();
        int[] lps = new int[sz];
        
        // KMP Preprocessing
        int j = 0;
        for (int i = 1; i < sz; i++) {
            while (j > 0 && part.charAt(i) != part.charAt(j)) {
                j = lps[j - 1];
            }
            if (part.charAt(i) == part.charAt(j)) {
                lps[i] = ++j;
            }
        }

        StringBuilder ans = new StringBuilder();
        int[] stk = new int[s.length()];
        j = 0;
        int idx = 0;

        for (char ch : s.toCharArray()) {
            ans.append(ch);
            stk[idx++] = j;

            while (j > 0 && ch != part.charAt(j)) {
                j = lps[j - 1];
            }

            if (ch == part.charAt(j)) {
                j++;
            }

            stk[idx - 1] = j;

            if (j == sz) {
                ans.setLength(ans.length() - sz);
                idx -= sz;
                j = (idx > 0) ? stk[idx - 1] : 0;
            }
        }

        return ans.toString();
    }
}
